src/query/functions/aggregation/base.go

```diff
@@ -18,58 +18,100 @@
 // THE SOFTWARE.
 
-package functions
+package aggregation
 
 import (
 	"fmt"
-	"math"
 
 	"github.com/m3db/m3/src/query/block"
 	"github.com/m3db/m3/src/query/executor/transform"
+	"github.com/m3db/m3/src/query/functions/utils"
 	"github.com/m3db/m3/src/query/parser"
 )
 
-// CountType counts all non nan elements in a list of series
-const CountType = "count"
+type aggregationFn func(values []float64, indices [][]int) []float64
 
-// CountOp stores required properties for count
-type CountOp struct {
+var aggregationFunctions = map[string]aggregationFn{
+	SumType:               sumFn,
+	MinType:               minFn,
+	MaxType:               maxFn,
+	AverageType:           averageFn,
+	StandardDeviationType: stddevFn,
+	StandardVarianceType:  varianceFn,
+	CountType:             countFn,
+}
+
+// NodeParams contains additional parameters required for aggregation ops
+type NodeParams struct {
+	Matching []string
+	Without  bool
+}
+
+// NewAggregationOp creates a new aggregation operation
+func NewAggregationOp(
+	opType string,
+	params NodeParams,
+) (parser.Params, error) {
+	if fn, ok := aggregationFunctions[opType]; ok {
+		return newBaseOp(params, opType, fn), nil
+	}
+	return baseOp{}, fmt.Errorf("operator not supported: %s", opType)
+}
+
+// baseOp stores required properties for count
+type baseOp struct {
+	params NodeParams
+	opType string
+	aggFn  aggregationFn
 }
 
 // OpType for the operator
-func (o CountOp) OpType() string {
-	return CountType
+func (o baseOp) OpType() string {
+	return o.opType
 }
 
 // String representation
-func (o CountOp) String() string {
+func (o baseOp) String() string {
 	return fmt.Sprintf("type: %s", o.OpType())
 }
 
 // Node creates an execution node
-func (o CountOp) Node(controller *transform.Controller) transform.OpNode {
-	return &CountNode{op: o, controller: controller}
+func (o baseOp) Node(controller *transform.Controller) transform.OpNode {
+	return &baseNode{
+		op:         o,
+		controller: controller,
+	}
 }
 
-// CountNode is an execution node
-type CountNode struct {
-	op         CountOp
+func newBaseOp(params NodeParams, opType string, aggFn aggregationFn) baseOp {
+	return baseOp{
+		params: params,
+		opType: opType,
+		aggFn:  aggFn,
+	}
+}
+
+type baseNode struct {
+	op         baseOp
 	controller *transform.Controller
 }
 
 // Process the block
-func (c *CountNode) Process(ID parser.NodeID, b block.Block) error {
-	// TODO: Figure out a good name and tags after an aggregation operation
-	meta := block.SeriesMeta{
-		Name: CountType,
-	}
-
+func (n *baseNode) Process(ID parser.NodeID, b block.Block) error {
 	stepIter, err := b.StepIter()
 	if err != nil {
 		return err
 	}
 
-	builder, err := c.controller.BlockBuilder(stepIter.Meta(), []block.SeriesMeta{meta})
+	params := n.op.params
+	indices, metas := utils.GroupSeries(
+		params.Matching,
+		params.Without,
+		n.op.opType,
+		stepIter.SeriesMeta(),
+	)
+
+	builder, err := n.controller.BlockBuilder(stepIter.Meta(), metas)
```

Comment:
shouldn't the metadata be different ? I think aggregation removes a lot of the metadata.